openlr/candidate_paths_getter.cpp

```diff
@@ -0,0 +1,348 @@
+  {
+    auto * start = l.get();
+    while (start->m_parent)
+      start = start->m_parent.get();
+
+    auto const bearStartPoint = pointsSelector.GetBearingStartPoint(start->m_edge);
+    auto const startPointsDsitance =
+        static_cast<uint32_t>(MercatorBounds::DistanceOnEarth(bearStartPoint, startPoint));
+
+    uint32_t traceBackIterationsLeft = 3;
+    for (auto const * part = &l; (*part); part = &((*part)->m_parent))
+    {
+      if (traceBackIterationsLeft == 0)
+        break;
+
+      --traceBackIterationsLeft;
+
+      auto const bearEndPoint =
+          pointsSelector.GetBearingEndPoint((*part)->m_edge, (*part)->m_distanceM);
+
+      auto const bearing = Bearing(bearStartPoint, bearEndPoint);
+      auto const bearingDiff = AbsDifference(bearing, requiredBearing);
+      auto const pathDistDiff = AbsDifference((*part)->m_distanceM, bearDistM);
+
+      // TODO(mgsergio): Check bearing is within tolerance. Add to canidates if it is.
+
+      if ((*part)->m_hasFake)
+        fakeEndingsCandidatePaths.emplace(*part, bearingDiff, pathDistDiff, startPointsDsitance);
+      else
+        candidatePaths.emplace(*part, bearingDiff, pathDistDiff, startPointsDsitance);
+      // candidatePaths.emplace(*part, bearingDiff, pathDistDiff, startPointsDsitance);
+    }
+  }
+
+  ASSERT(
+      none_of(begin(candidatePaths), end(candidatePaths), mem_fn(&CandidatePath::HasFakeEndings)),
+      ());
+  ASSERT(fakeEndingsCandidatePaths.empty() ||
+             any_of(begin(fakeEndingsCandidatePaths), end(fakeEndingsCandidatePaths),
+                    mem_fn(&CandidatePath::HasFakeEndings)),
+         ());
+
+  vector<CandidatePath> paths;
+  copy_n(begin(candidatePaths), min(static_cast<size_t>(kMaxCandidates), candidatePaths.size()),
+         back_inserter(paths));
+
+  copy_n(begin(fakeEndingsCandidatePaths),
+         min(static_cast<size_t>(kMaxFakeCandidates), fakeEndingsCandidatePaths.size()),
+         back_inserter(paths));
+
+  LOG(LDEBUG, ("List candidate paths..."));
+  for (auto const & path : paths)
+  {
+    LOG(LDEBUG, ("CP:", path.m_bearingDiff, path.m_pathDistanceDiff, path.m_startPointDistance));
+    Graph::EdgeVector edges;
+    for (auto * p = path.m_path.get(); p; p = p->m_parent.get())
+      edges.push_back(p->m_edge);
+    if (!isLastPoint)
+      reverse(begin(edges), end(edges));
+
+    candidates.emplace_back(move(edges));
+  }
+}
+
+void CandidatePathsGetter::GetLineCandidates(openlr::LocationReferencePoint const & p,
+                                             bool const isLastPoint,
+                                             uint32_t const distanceToNextPoint,
+                                             vector<m2::PointD> const & pointCandidates,
+                                             vector<Graph::EdgeVector> & candidates)
+{
+  ScopedTimer t(m_stat.m_lineCandidatesTime);
+
+  uint32_t const kBearTolerance = 30;
+
+  uint32_t const bearDistM = std::min(25u, distanceToNextPoint);
```

Comment:
Please, create a named constant for 25u?

Reply:
Done.